ngraph/core/src/preprocess/pre_post_process.cpp

```diff
@@ -346,9 +444,73 @@ std::shared_ptr<Function> PrePostProcessor::build(const std::shared_ptr<Function
     }
+
+    // Validate nodes after preprocessing if needed (no need to repeat it after post-processing)
     if (tensor_data_updated) {
         function->validate_nodes_and_infer_types();
     }
+
+    // Post processing
+    for (const auto& output : m_impl->out_contexts) {
+        std::shared_ptr<op::v0::Result> result;
+        Output<Node> node;
+        OPENVINO_ASSERT(output, "Internal error: Invalid postprocessing output, please report a problem");
+        if (output->has_index()) {
+            node = function->output(output->m_index);
+        } else if (output->has_name()) {
+            node = function->output(output->m_name);
+        } else {
+            node = function->output();
+        }
+        result = std::dynamic_pointer_cast<op::v0::Result>(node.get_node_shared_ptr());
+        // Set result layout from 'network' information
+        if (output->m_network_data && output->m_network_data->is_layout_set() && result->get_layout().empty()) {
+            result->set_layout(output->m_network_data->get_layout());
+        }
+        auto parent = result->get_input_source_output(0);
+        if (!output->m_tensor_data) {
+            output->create_tensor_data();
+        }
+        PostprocessingContext context(result->get_layout());
+        if (output->m_tensor_data->is_layout_set()) {
+            context.target_layout() = output->m_tensor_data->get_layout();
+        }
+        if (output->m_tensor_data->is_element_type_set()) {
+            context.target_element_type() = output->m_tensor_data->get_element_type();
+        }
+        // Apply post-processing
+        node = result->get_input_source_output(0);
+        if (output->m_postprocess) {
+            for (const auto& action : output->m_postprocess->actions()) {
+                auto action_result = action({node}, context);
+                node = std::get<0>(action_result);
+            }
+        }
+        // Implicit: Convert element type + layout to user's tensor implicitly
+        PostStepsList implicit_steps;
+        if (node.get_element_type() != output->m_tensor_data->get_element_type() &&
+            output->m_tensor_data->is_element_type_set() && node.get_element_type() != element::dynamic) {
+            implicit_steps.add_convert_impl(output->m_tensor_data->get_element_type());
+        }
+
+        if (!context.target_layout().empty() && context.target_layout() != context.layout()) {
+            implicit_steps.add_convert_layout_impl(context.target_layout());
+        }
+        for (const auto& action : implicit_steps.actions()) {
+            auto action_result = action({node}, context);
+            node = std::get<0>(action_result);
+        }
+
+        // Create result
+        auto new_result = std::make_shared<ov::op::v0::Result>(node);
+        if (!context.layout().empty()) {
+            new_result->set_layout(context.layout());
+        }
+        new_result->get_input_tensor(0).set_names(result->get_input_tensor(0).get_names());
```

Comment:
should not we reset tensor names for the previous / original tensor?

How is it done for inputs?